### DISCUSSION TITLE
Feature request: efficient PSD flattening

### DISCUSSION THREAD

**Discussion by jonsneyers** [ImageMagick/ImageMagick#6714] [Contribute]
Some PSD files consist of several layers, but they don't store the final merged image explicitly. In those cases, the PSD decoder currently first decodes all layers, and then merges them. This could be optimized in terms of memory usage by interleaving the merging with the decoding, so only 2 buffers are needed instead of N+1.

**Comment by dlemstra** [comment]
Is this the case where you only want to read the merged image? We could probably add an optimization when we read the layers for this. Is it possible to share an image that we can use to implement this?

**Comment by jonsneyers** [comment]
[OpacityFunNoMerged.psd.zip](https://github.com/ImageMagick/ImageMagick/files/2648173/OpacityFunNoMerged.psd.zip)
![opacityfun](https://user-images.githubusercontent.com/15010068/49504956-05fe0700-f87b-11e8-8fe3-fe4d70ce359b.png)

Here is a test image which has some layers but no merged image.
The PNG file is what it should look like (this is what you get if you export it to PNG in Photoshop).
Note that it looks like both Gimp and ImageMagick are not decoding it correctly (the black rectangle is missing), so I will open another issue for that.

**Comment by paperboyo** [comment]
Just a comment. For PSDs, exiftool reports presence of a merged composite via `exiftool -HasRealMergedData ` (`0x0421`). Sadly for layered TIFF files, one needs to look for [`MTrn`, `Mt16` or `Mt32`](https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#50577409_pgfId-1049436) via `exiftool -v` (`HasRealMergedData` is, ~wrongly!,~ always ON for TIFFs). I don’t think ImageMagick surfaces this metadata.

This particular file above can be safely merged via ImageMagick. But PSDs and TIFFs without saved composite (and for TIFFs this is [the default](https://feedback-readonly.photoshop.com/conversations/photoshop/photoshop-default-tiff-save-options-to-be-save-transparency-on/5f5f46104b561a3d42719a47)!) aren‘t really always safe for merging. One would need to painstakingly create a whitelist of features that ImageMagick can reliable composite and error or warn if the files contain fancy/exotic dynamic features (layer effects, gradient layers, unsupported blend modes etc). I don’t think all of them contain pixel copies (like Type layers here). Luckily, it looks like the structure of the data is the more or less the same for PSD and layered TIFFs. 

There is another serious problem for when a TIFF file contains a loose Alpha channel(s) (haven’t checked PSD yet) even with a composite merged layer. ImageMagick interprets its `Unspecified` alpha as transparency and burns holes where Photoshop didn’t mean to. 

I only commented, because I’m on a arduous mission to try to come up with a logic of where it is safe to convert layered TIFFs via ImageMagick, where some need to be treated differently and where the app I’m using will have to say _sorry, but I can’t reliably say I can support it_. This is [my current thinking](https://github.com/guardian/grid/issues/2395#issuecomment-939602668) (unfinished). If anyone has any ideas of where this could be simplified, I would love to hear!

Related: https://github.com/ImageMagick/ImageMagick/issues/1315.

**Comment by fmw42** [comment]
FYI, Imagemagick, to my knowledge, does not currently handle layered TIFFs, only TIFFs with multiple pages.

**Comment by paperboyo** [comment]
I mean, it does. Just not always 😜. And it’s very, very hard to even say upfront when it’s likely to go wrong (which would be a great start). I’m mostly (if not exclusively) after using merged result. Even though there is access to individual layers, I don’t much care for those. I care even less about writing them out.
My problem is that simply converting a layered TIFF to a PNG can go horribly wrong and is unpredictable. Being able to tell when it’s not going to be done correctly (sometimes when there is no composite image*, it’s just practically impossible, unless Adobe opensources their compositing engine?) is the first step in supporting layered TIFFs and PSDs. Second step would be to handle all the cases of dealing with the merged result that can actually be handled – correctly. Only last step would be supporting interchange of layers with other layered file formats, in my humblest of opinions.

*[EDIT] for PSD there may not be a composite image at all (false `HasRealMergedData`), TIFF always has one, but the transparency of it may be gone (Save Transparency checkbox off)

**Comment by paperboyo** [comment]
> does not currently handle layered TIFFs

Just for fun, this was my weekend trying to come up with some [kind of logic](https://github.com/guardian/grid/issues/2395#issuecomment-939602668) where stuff can be more or less predictable (again, only while converting to PNG with/without transparency):

![image](https://user-images.githubusercontent.com/6032869/136847181-47ed44dd-8be7-4903-9ba7-60a1bf24311d.png)

Green works out of the box. Red may never work (so needs identification). Orange needs identification and its own special path. The real pain is to tell them colours apart on arrival…

**Comment by fmw42** [comment]
Photoshop PSD and TIFF layered files, especially PSD files are quite complex because they have many features. The IM developers have day jobs and so are always very busy. So I do not think it is high priority to fully support all aspects of these files. Individual features may be considered from time to time depending upon time availability. Bugs are always fixed first. Any user is welcome to contribute enhancements.
